### PR TITLE
Ledger Support & Ledger Dev Tools

### PR DESCRIPTION
**Pull request by dawidsowardx**
* Add DevTools 
  * Wallet Simulator
  * LedgerSimulator (generating valid WebRTC responses based on "real world" inputs)
* Communication with Ledger device
  * Get Device ID
  * Get Device Model
  * Get Public Key (both curves)

### CODE REVIEW

File: src/ledger/ledger-radix-babylon.ts [hunk 1]
```diff
@@ -0,0 +1,87 @@
...
+  const parts = path
+    .map((value) => 0x80000000 + parseInt(value, 10))
+    .map((value) => value.toString(16))
+    .join('')
+
+  return `${length}${parts}`
+}
+
+export enum LedgerErrorResponse {
+  UnlockDevice = '5515',
```
Comment (xstelea) [inline src/ledger/ledger-radix-babylon.ts:18]: is it intended to mix string and number values?

File: src/ledger/ledger-radix-babylon.ts [hunk 2]
```diff
@@ -0,0 +1,87 @@
...
+  OpenRadixApp = 28161,
+}
+
+export enum LedgerInstructionCode {
```
Comment (xstelea) [inline src/ledger/ledger-radix-babylon.ts:22]: we're not using enums as a code convention. Lets make a decision before introduction them

File: src/ledger/ledger-radix-babylon.ts [hunk 3]
```diff
@@ -0,0 +1,87 @@
...
+  GetDeviceModel = 0x11,
+  GetDeviceId = 0x12,
+  GetPubKeyEd25519 = 0x21,
+  GetPubKeySecp256k1 = 0x31,
+}
+
+export const errorResponses: Record<string, string> = {
```
Comment (xstelea) [inline src/ledger/ledger-radix-babylon.ts:29]: `Record<LedgerErrorResponse, string>`

File: src/ledger/ledger-radix-babylon.ts [hunk 4]
```diff
@@ -0,0 +1,87 @@
...
+  [LedgerErrorResponse.UnlockDevice]:
+    'Please unlock Ledger Device and try again',
+  [LedgerErrorResponse.OpenRadixApp]:
+    'Please open Radix Babylon app in your Ledger device and try again',
+}
+
+export class LedgerRadixBabylon {
```
Comment (xstelea) [inline src/ledger/ledger-radix-babylon.ts:36]: Is there a specific reason why you are using a class? Same functionality can be accomplished with a function and not introduce classes into the codebase

File: src/ledger/ledger-radix-babylon.ts [hunk 5]
```diff
@@ -0,0 +1,87 @@
...
+      ),
+      (err: any) => {
+        if (err?.name === 'LockedDeviceError') {
+          return errorResponses[LedgerErrorResponse.UnlockDevice]
+        }
+
+        return errorResponses[err?.statusCode] || err.message
+      }
+    ).andThen((buffer) =>
+      ok(buffer.slice(0, buffer.length - 2).toString('hex'))
```
Comment (xstelea) [inline src/ledger/ledger-radix-babylon.ts:84]: `The signature '(start?: number | undefined, end?: number | undefined): Buffer' of 'buffer.slice' is deprecated.`

File: src/offscreen/create-offscreen.ts [hunk 6]
```diff
@@ -0,0 +1,12 @@
+import { config } from 'config'
+
+export const createOffscreen = async () => {
+  if (await chrome.offscreen.hasDocument()) {
+    await chrome.offscreen.closeDocument()
+  }
+  await chrome.offscreen.createDocument({
+    url: config.offscreen.url,
+    reasons: [chrome.offscreen.Reason.WEB_RTC],
```
Comment (xstelea) [inline src/offscreen/create-offscreen.ts:9]: and websockets?

File: src/offscreen/offscreen.ts [hunk 7]
```diff
@@ -0,0 +1,42 @@
...
+connectorClient.onMessage$.subscribe(chrome.runtime.sendMessage)
+
+declare global {
+  // eslint-disable-next-line @typescript-eslint/consistent-type-definitions
+  interface Window {
+    sendWebRTCMessage: typeof connectorClient.sendMessage
+  }
+}
+
+window.sendWebRTCMessage = connectorClient.sendMessage
```
Comment (xstelea) [inline src/offscreen/offscreen.ts:23]: why do we want to add this to the global scope?

File: src/offscreen/offscreen.ts [hunk 8]
```diff
@@ -0,0 +1,42 @@
...
+        Buffer.from(message.connectionPassword, 'hex')
+      )
+      connectorClient.connect()
+    } else {
+      connectorClient.disconnect()
+    }
+  }
+})
+
+chrome.runtime.sendMessage('offscreen-document-ready')
```
Comment (xstelea) [inline src/offscreen/offscreen.ts:42]: We should have a unified message format for our own sanity
something like: 
```
{
  discriminator: 'offscreenDocumentReady'
}
```

**Review by xstelea — CHANGES_REQUESTED** [review]

File: src/offscreen/create-offscreen.ts (hunk 6 above)
Comment (dawidsowardx) [reply src/offscreen/create-offscreen.ts:9]: Websockets are not listed in reason enum so I don't think we can/need to include it here

File: src/offscreen/offscreen.ts (hunk 7 above)
Comment (dawidsowardx) [reply src/offscreen/offscreen.ts]: reverted

File: src/ledger/ledger-radix-babylon.ts (hunk 4 above)
Comment (dawidsowardx) [reply src/ledger/ledger-radix-babylon.ts]: I can't find the article but all ledger-live apps like [BTC](https://github.com/LedgerHQ/ledger-live/tree/develop/libs/ledgerjs/packages/hw-app-btc) and [ETH](https://github.com/LedgerHQ/ledger-live/tree/develop/libs/ledgerjs/packages/hw-app-eth) implement same interface I wanted to follow. Later I realized we don't need it so moving back to function and closure for now :)

File: package.json [hunk 9]
```diff
@@ -30,9 +30,16 @@
     "test:watch": "jest --watch"
   },
   "dependencies": {
+    "@ledgerhq/hw-transport-webhid": "^6.27.12",
     "@stitches/react": "^1.2.8",
+    "@types/blake2b": "^2.1.0",
     "bech32": "^2.0.0",
+    "bip32": "^2.0.0",
```
Comment (xstelea) [inline package.json:37]: this one is 2 major versions behind

**Review by xstelea — CHANGES_REQUESTED** [review]

File: src/chrome/background/message-handler.ts [hunk 10]
```diff
@@ -66,6 +70,30 @@ export const BackgroundMessageHandler =
         }))
       }
 
+      case messageDiscriminator.walletToLedger:
+        return createOrFocusTab(config.popup.pages.ledger)
+          .map((tab) => {
```
Comment (xstelea) [inline src/chrome/background/message-handler.ts:75]: change to `andThen` you are returning a ResultAsync through `sendMessageWithConfirmation`

File: src/chrome/dev-tools/components/LedgerSimulator.tsx [hunk 11]
```diff
@@ -0,0 +1,213 @@
...
+import blake2b from 'blake2b'
+import { logger } from 'utils/logger'
+import { sendMessage } from 'chrome/messages/send-message'
+import { createMessage } from 'chrome/messages/create-message'
+import { compiledTxHex } from '../example'
+
+const secp256k1 = new Elliptic('secp256k1')
+const ed25519 = new Elliptic('ed25519')
+
+const buf2hex = (buffer: any) =>
```
Comment (xstelea) [inline src/chrome/dev-tools/components/LedgerSimulator.tsx:21]: Just do this instead -> `buffer.toString('hex')`

File: src/chrome/offscreen/message-handler.ts [hunk 12]
```diff
@@ -11,24 +11,29 @@ import {
...
   const messageRouter = input.messageRouter
-  const logger = input.logger
+  const logger = input.logger || appLogger
 
   return (
     message: Message,
     sendMessageWithConfirmation: SendMessageWithConfirmation,
     tabId?: number
   ): MessageHandlerOutput => {
+    logger.debug('offscreen message handler', message)
```
Comment (xstelea) [inline src/chrome/offscreen/message-handler.ts:36]: this log is going to get noisy and not really reflect the messages that are handled

**Review by xstelea — APPROVED** [review]
LGTM

**Pull request merged**